src/crypter.rs

```diff
@@ -0,0 +1,145 @@
+};
+use base64::{engine::general_purpose::STANDARD, DecodeError, Engine};
+use std::{
+    collections::VecDeque,
+    fmt::{self, Debug, Formatter},
+    iter,
+    str::FromStr,
+    sync::Arc,
+};
+use typenum::marker_traits::Unsigned;
+
+pub type Key = aes_gcm::Key<AesGcm>;
+
+#[derive(Clone)]
+pub struct Crypter(Arc<CrypterInner>);
+
+#[derive(thiserror::Error, Debug, Clone)]
+pub enum CrypterParseError {
+    #[error(transparent)]
+    Base64(#[from] DecodeError),
+
+    #[error("incorrect key length, must be {} bytes after base64 decode", <AesGcm as KeySizeUser>::KeySize::USIZE)]
+    IncorrectLength,
+
+    #[error("at least one key needed")]
+    Missing,
+}
+
+impl FromStr for Crypter {
+    type Err = CrypterParseError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut keys = s
+            .split(',')
+            .map(|s| {
+                STANDARD
+                    .decode(s)
+                    .map_err(CrypterParseError::Base64)
+                    .and_then(|v| Key::from_exact_iter(v).ok_or(CrypterParseError::IncorrectLength))
+            })
+            .collect::<Result<VecDeque<Key>, _>>()?;
+        let current_key = keys.pop_front().ok_or(CrypterParseError::Missing)?;
+        Ok(Self::new(current_key, keys))
+    }
+}
+
+impl Debug for Crypter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Crypter")
+            .field("current_key", &"..")
+            .field("past_keys", &self.0.past_keys.len())
+            .finish()
+    }
+}
+
+#[derive(Clone, Debug)]
+struct CrypterInner {
+    current_key: Key,
+    past_keys: Vec<Key>,
+}
+
+impl From<Key> for Crypter {
+    fn from(key: Key) -> Self {
+        Self::new(key, [])
+    }
+}
+
+impl Crypter {
+    pub fn new(current_key: Key, past_keys: impl IntoIterator<Item = Key>) -> Self {
+        Self(Arc::new(CrypterInner {
+            current_key,
+            past_keys: past_keys.into_iter().collect(),
+        }))
+    }
+
+    pub fn random() -> Self {
+        Self::generate_key().into()
+    }
+
+    pub fn generate_key() -> Key {
+        AesGcm::generate_key(OsRng)
+    }
+
+    pub fn encrypt(&self, associated_data: &[u8], plaintext: &[u8]) -> Result<Vec<u8>, Error> {
+        self.0.encrypt(associated_data, plaintext)
+    }
+
+    pub fn decrypt(
+        &self,
+        associated_data: &[u8],
+        nonce_and_ciphertext: &[u8],
+    ) -> Result<Vec<u8>, Error> {
+        self.0.decrypt(associated_data, nonce_and_ciphertext)
+    }
+}
+
+impl CrypterInner {
+    fn encrypt(&self, associated_data: &[u8], plaintext: &[u8]) -> Result<Vec<u8>, Error> {
+        let nonce = AesGcm::generate_nonce(&mut OsRng);
+        let mut buffer = plaintext.to_vec();
+        AesGcm::new(&self.current_key).encrypt_in_place(&nonce, associated_data, &mut buffer)?;
+        let mut nonce_and_ciphertext = nonce.to_vec();
+        nonce_and_ciphertext.append(&mut buffer);
```

Comment:
Storing the nonce first, and then the AES-GCM ciphertext is fine. Since the nonce is of a fixed size, it's unambiguously decodable. In fact, this is the more conventional storage order. (it is marginally more cache-sympathetic too)